### ISSUE TITLE
llvm assertion failed: Failed to convert memory constraint code to constraint id in `std.atomic.Atomic bitRmw()`

### ISSUE THREAD

**Issue by mikdusan** [ziglang/zig#9163]
- known to manifest on macos, linux, dragonfly
- known to manifest on llvm 12.0.0.1
- known to manifest on llvm 12.0.1-rc2

---

- `reduction.zig`
```zig
const std = @import("std");

test "" {
    var x = std.atomic.Atomic(u32).init(0);
    _ = x.bitSet(3, .Monotonic);
}
```

```sh
$ sw_vers
ProductName:	macOS
ProductVersion:	11.4
BuildVersion:	20F71
$ /opt/llvm-macos11.0-x86_64-12.0.1-rc2-release-assert/bin/llvm-config  --build-mode --assertion-mode
Release
ON
$ zig test reduction.zig
Assertion failed: (ConstraintID != InlineAsm::Constraint_Unknown && "Failed to convert memory constraint code to constraint id."), function visitInlineAsm, file /opt/src/llvm/12.0.1-rc2/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp, line 8511.
zsh: abort      zig test reduction.zig
```

---

see this asm segment:
https://github.com/ziglang/zig/blob/e006281d79a31e10bdbc07b3e33427d719ec01df/lib/std/atomic/Atomic.zig#L194-L199

**Comment by LemonBoy** [comment]
cc @kprotty The `p` constraint is not supported by LLVM, replacing it with `m` solves the problem and should not generate invalid asm (I can only guess the `p` was added to ensure the pointer is a valid memory operand).

**Comment by kprotty** [comment]
@LemonBoy I've been using the `p` constraint for a while now with [zig-adaptive-lock](https://github.com/kprotty/zig-adaptive-lock/tree/std_lock) and it seems to pass CI. Is it an LLVM 12 thing? Don't remember exactly but I think the reason I switched to `p` over `m` was due to some description of gcc inline asm saying it was more versatile or something. If it doesn't affect codegen, would be fine to convert it back to `m`.

**Comment by LemonBoy** [comment]
The check's only present if LLVM is built in debug mode, for some reason the unimplemented `p` falls trough every error check and produces the equivalent of `m`.

>  I think the reason I switched to p over m was due to some description of gcc inline asm saying it was more versatile or something

Use LLVM's [langref](https://llvm.org/docs/LangRef.html#inline-asm-constraint-string) as reference, many "exotic" inline-asm constraints aren't implemented yet.

**Comment by Vexu** [comment]
Closed by #11523
